src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/containers/ContainerManager.java

```diff
@@ -87,170 +105,327 @@ public void setTimeoutMultiplier(int multiplier) {
     containerBackend.setTimeoutMultiplier(multiplier);
   }
 
+  private static Pattern bindingRegex = Pattern.compile("@([a-zA-Z_][-a-zA-Z_0-9]*)@");
+
+  private ExecutionConfig.Builder applyBindings(String command,
+                                                ImmutableMap<String, Binding> bindings,
+                                                Map<String, ContainerExecResponse> stepResults,
+                                                File containerTempDir)
+      throws ContainerExecutionException, IOException {
+    ExecutionConfig.Builder builder = ExecutionConfig.builder();
+    StringBuilder finalCommand = new StringBuilder();
+
+    Map<String, Binding> mutableBindings = new HashMap<String, Binding>(bindings);
+
+    Matcher regexMatcher = bindingRegex.matcher(command);
+    while (regexMatcher.find()) {
+      String name = regexMatcher.group(1);
+      Binding binding;
+      if (mutableBindings.containsKey(name)) {
+        binding = mutableBindings.get(name);
+      } else if (stepResults.containsKey(name)) {
+        File stepFile = new File(containerTempDir, name);
+        try (FileWriter w = new FileWriter(stepFile)) {
+          w.write(stepResults.get(name).response());
+        }
+        binding = new FileBinding(stepFile, false);
+        mutableBindings.put(name, binding);
+      } else {
+        throw new ContainerExecutionException("Couldn't find a binding called " + name
+            + " for command " + command);
+      }
+      if (binding.needsApplying()) {
+        binding.applyBinding(builder, name);
+      }
+      regexMatcher.appendReplacement(finalCommand, binding.getMountPoint(name));
+    }
+    regexMatcher.appendTail(finalCommand);
+
+    Matcher matcher = COMMAND_TOKENIZER.matcher(finalCommand);
+    while (matcher.find()) {
+      builder.addCommand(matcher.group());
+    }
+    return builder;
+  }
+
+  abstract static class Binding {
+    abstract String getMountPoint(String name);
+
+    ExecutionConfig.Builder applyBinding(ExecutionConfig.Builder builder, String name) {
+      throw new UnsupportedOperationException();
+    }
+
+    boolean needsApplying() {
+      return false;
+    }
+  }
+
+  static class FileBinding extends Binding {
+    private final File file;
+    private final boolean readWrite;
+    private boolean needsApplying;
+
+    FileBinding(File file, boolean readWrite) {
+      this.file = file;
+      this.readWrite = readWrite;
+      this.needsApplying = true;
+    }
+
+    @Override
+    ExecutionConfig.Builder applyBinding(ExecutionConfig.Builder builder, String name) {
+      needsApplying = false;
+      return builder.addPathSpecification(
+          PathSpecification.create(file, getMountPoint(name), readWrite));
+    }
+
+    @Override
+    String getMountPoint(String name) {
+      return "/mnt/pottery/" + name;
```

Comment:
I think so; since it's internal to the container, I've put it on the `ContainerBackend` service; `ContainerEnvConfig` seemed to be things external to the container.